Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -127,28 +129,32 @@ end Polar
 /-- A quadratic form over a module.
 
 For a more familiar constructor when `R` is a ring, see `QuadraticForm.ofPolar`. -/
-structure QuadraticForm (R : Type u) (M : Type v) [Semiring R] [AddCommMonoid M] [Module R M] where
-  toFun : M → R
-  toFun_smul : ∀ (a : R) (x : M), toFun (a • x) = a * a * toFun x
-  exists_companion' : ∃ B : BilinForm R M, ∀ x y, toFun (x + y) = toFun x + toFun y + B x y
+structure QuadraticForm (R : Type u) (M : Type v) (N : Type w) [Semiring R] [AddCommMonoid M]
+    [Module R M] [Module Rᵐᵒᵖ M] [AddCommMonoid N] [Module R N] [Module Rᵐᵒᵖ N]
+    [SMulCommClass Rᵐᵒᵖ R N] where
+  toFun : M → N
+  toFun_smul : ∀ (a : R) (x : M), toFun (a • x) = (a * a) • toFun x
+  toFun_smulr : ∀ (a : Rᵐᵒᵖ) (x : M), toFun (a • x) = (a * a) • toFun x
```

Comment:
This feels a bit like this is trying to be a quadratic map over a bimodule, which IMO is a little awkward given that we don't even have linear maps over a bimodule!

Do you actually need to support a non-commutative `R` here? I don't know if even the current definition makes any sense when R is not commutative.

Reply:
> This feels a bit like this is trying to be a quadratic map over a bimodule, which IMO is a little awkward given that we don't even have linear maps over a bimodule!

Yes, I got the idea for how to do this from https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Algebra/Module/Bimodule.lean

> Do you actually need to support a non-commutative `R` here? I don't know if even the current definition makes any sense when R is not commutative.

Personally I'd be quite happy with `R` commutative. The current `QuadraticForm` definition uses `R` non-commutative. So the possibilities are:

1. Develop `QuadraticMap` and `QuadraticForm` as separate (but very similar) theories, with the former over a commutative ring and the latter over a non-commutative ring
2. Have `QuadraticForm` as a special case of `QuadraticMap`, but both over a commutative ring
3. Figure out how to make `QuadraticForm` work over a non-commutative ring

I guess there is also the option of starting off with `QuadraticForm` work over a non-commutative ring and then specialising to commutative when we need to?

Reply:
I'm starting to like the idea of 2. Maybe that looks like:
* make a PR that ungeneralizes QuadraticForm to commutative rings and see if CI/Zulip cares
* generalize `QuadraticForm` to `QuadraticMap`.

Reply:
It might be worth asking on Zulip and looking at the git history before embarking on that first point

Reply:
Looks like the use of non-commutative rings goes back to the PR that originally introduced quadratic forms:

https://github.com/leanprover-community/mathlib/pull/2480

Reply:
I've asked here: https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Quadratic.20Maps